### ISSUE TITLE
Hierarchical allocation tree: nested sub-portfolios with per-sleeve budgets, strategies, and weighting

### ISSUE THREAD

**Issue by jugegp16** [jugegp16/Midas#33]
## Background

Midas's allocator currently treats the portfolio as a flat list of tickers with global constraints (`max_position_pct`, `min_cash_pct`, `sigmoid_steepness`, etc.). One strategy mix applies to all tickers. One cap applies to all tickers. One weighting policy applies to all tickers.

This is the wrong abstraction for portfolios that mix conceptually distinct sleeves. Looking at the actual `portfolio.yaml` (21 tickers):

- **Broad-market core**: VOO, SCHX, SCHA, QQQM — should be low-turnover, low-conviction-tilt, mostly held passively.
- **Mega-cap tech**: NVDA, MSFT, AAPL, META, ADBE, SMH — actively traded on momentum, high vol, high correlation among themselves.
- **Sector tilts**: XLU, VDE, VFH, CEG, SLB — sector rotation, mean-reversion friendly, want bounded individual exposure.
- **Speculative growth**: PLTR, LCID, FRMI, INCY — high vol, high conviction needed, hard cap on individual exposure regardless of how bullish a strategy gets.
- **Crypto**: ETH, IBIT — separate asset class entirely, want risk-parity weighting within the sleeve and a hard cap on total crypto exposure.

Right now all 21 tickers compete in one softmax with one set of strategies and one max_position_pct. There is no way to express "speculative names get capped at 5% individually but momentum can still be aggressive on tech mega-caps" or "crypto exposure tops out at 5% of total portfolio regardless of softmax math."

## Proposal

Model the portfolio as a tree of nested sub-allocators. Each node has:

- A **budget** (fraction of parent's budget)
- A list of **tickers** OR a list of **child nodes**
- Its own **strategies** (CONVICTION + PROTECTIVE)
- Its own **weighting policy** (`equal`, `inverse_vol`, etc. — see #32)
- Its own **constraints** (`max_position_pct`, `rebalance_threshold`)

Example config:

```yaml
portfolio:
  budget: 0.95              # investable fraction of total
  weighting: equal          # how children's budgets relate to each other
  children:
    - name: us_equity_core
      budget: 0.45          # 45% of root = 42.75% of total portfolio
      weighting: equal
      tickers: [VOO, SCHX, SCHA, QQQM]
      strategies:
        conviction:
          - { name: Momentum, weight: 1.0, params: { window: 60 } }

    - name: tech_concentration
      budget: 0.25
      weighting: inverse_vol
      max_position_pct: 0.10
      tickers: [NVDA, MSFT, AAPL, META, ADBE, SMH]
      strategies:
        conviction:
          - { name: Momentum, weight: 2.0 }
          - { name: ProfitTaking, weight: 1.0 }

    - name: sector_tilts
      budget: 0.15
      weighting: equal
      tickers: [XLU, VDE, VFH, CEG, SLB]
      strategies:
        conviction:
          - { name: Momentum, weight: 1.0 }
          - { name: MeanReversion, weight: 1.0 }

    - name: speculative
      budget: 0.10
      weighting: equal
      max_position_pct: 0.04   # hard cap inside sleeve
      tickers: [PLTR, LCID, FRMI, INCY]
      strategies:
        conviction:
          - { name: Momentum, weight: 1.5 }
        protective:
          - { name: TrailingStop, weight: 1.0, veto_threshold: -0.3 }

    - name: crypto
      budget: 0.05
      weighting: inverse_vol
      max_position_pct: 0.03
      tickers: [ETH, IBIT]
      strategies:
        conviction:
          - { name: Momentum, weight: 1.0 }
        protective:
          - { name: TrailingStop, weight: 1.0, veto_threshold: -0.4 }
```

Each sleeve runs an independent softmax over its own ticker universe with its own strategies. Sleeve budgets are guaranteed by construction — no amount of bullish conviction in `tech_concentration` can spill over and crowd out the `crypto` budget. Per-sleeve `max_position_pct` lets you bound speculative exposure at 4% per name without forcing the same cap on tech mega-caps.

This is the standard pattern in:

- **Composer** (composer.trade) — strategies are tree-structured "symphonies" with weighted nodes.
- **Wealthfront** — taxable + retirement + bond + international portfolios are tree nodes with per-node rebalancing.
- **Betterment** — stock/bond glide path is a tree with per-asset-class allocations.
- **QuantConnect LEAN** — `Universe`-scoped portfolio construction models compose into a global construction layer.
- **Riskfolio-Lib `HRP` (Hierarchical Risk Parity)** — recursive bisection over a correlation-clustering tree.
- **Bloomberg PORT, MSCI Barra, Axioma** — institutional optimizers all support nested constraints by sleeve.

## Why this comes after #32 (pluggable weighting)

The tree node's `weighting` field is the same pluggable policy from #32. Without that infrastructure in place, every tree node would be hard-coded to one weighting scheme. With #32 done first, the tree is a thin orchestration layer over composable allocator primitives.

Order of work:
1. #32 — pluggable weighting + drawdown scaling
2. This issue — hierarchical tree over the existing allocator

## Architectural sketch

New types:

```python
@dataclass
class AllocationNode:
    name: str
    budget: float                          # fraction of parent
    weighting: WeightingPolicy
    constraints: AllocationConstraints     # local override of parent
    # exactly one of:
    tickers: list[str] | None
    children: list["AllocationNode"] | None
    strategies: list[ScoredStrategy]
    protective: list[ProtectiveEntry]

class PortfolioAllocator:
    """Top-level allocator that owns a tree of AllocationNodes."""
    def allocate(self, ...) -> AllocationResult:
        # Recursively allocate budgets to nodes, then to leaves.
```

The existing `Allocator` class becomes the leaf-node primitive. `PortfolioAllocator` is a new orchestrator that walks the tree top-down: at each internal node, distribute the parent's budget across children (by their declared `budget` fractions, possibly with their own softmax over child-level scores in a future extension); at each leaf, run the existing softmax allocator over that leaf's ticker universe with that leaf's budget, strategies, and weighting.

`AllocationResult` extends to include sleeve attribution: every target carries its sleeve path (`tech_concentration/NVDA`) so the rebalancer can label trades with their sleeve. The strategy attribution from the rebalancer remains unchanged.

Backwards compatibility: a flat portfolio config (current `portfolio.yaml` shape) is treated as a single root node containing all tickers — the tree degenerates to the current allocator. Existing YAMLs work without modification.

## Cap interaction

Per-sleeve caps (`speculative.max_position_pct: 0.04`) are enforced inside the sleeve's softmax via the existing cap-with-redistribution logic from #29. Cross-sleeve caps don't exist by construction — the sleeve budget IS the cross-sleeve cap. Want to limit total crypto to 5%? Set the crypto sleeve's budget to 0.05.

## Strategy interaction

Strategies are scoped to their containing sleeve. `Momentum` in `tech_concentration` only sees NVDA/MSFT/AAPL/etc. — it never scores VOO. This solves another latent problem: today, every strategy sees every ticker, which means strategy parameters that work great for high-vol tech ruin the optimizer's signal on broad-market ETFs (and vice versa). Sleeve-scoped strategies let the optimizer tune `Momentum.window=20` for tech and `Momentum.window=60` for broad-market core simultaneously.

This also enables the optimizer to tune sleeve-specific strategy weights independently — a much richer search space than the current "one strategy mix for everything."

## Scope (rough)

- New types: `AllocationNode`, `PortfolioAllocator`, sleeve-aware `AllocationResult`. ~150 lines.
- Recursive allocation walker in `PortfolioAllocator.allocate()`. ~80 lines.
- Config parser: `src/midas/config.py` learns to parse nested `children`. ~50 lines.
- Backtest/live integration: pass `PortfolioAllocator` instead of `Allocator`. ~20 lines.
- Optimizer integration: search space gains per-sleeve hyperparameters. ~100 lines.
- Tests: tree allocation, sleeve budgets enforced, per-sleeve caps, sleeve-scoped strategies, backwards compat with flat config. ~250 lines.
- `docs/architecture.md`: new section on hierarchical allocation. ~50 lines.
- Migration: convert `sample-portfolio.yaml` and `portfolio.yaml` to tree form (or keep flat as backwards-compat examples). Author a tree version of `portfolio.yaml` showing the proposed sleeve structure.

Estimated 600-800 lines total. This is a real architectural PR — worth a design review before implementation.

## Success criteria

- [ ] Flat portfolio config (current shape) continues to work without modification
- [ ] Tree config with nested children allocates correctly per sleeve
- [ ] Per-sleeve `max_position_pct` is enforced inside the sleeve, doesn't bleed across sleeves
- [ ] Sleeve budgets are guaranteed: bullish conviction in one sleeve cannot spill into another's budget
- [ ] Sleeve-scoped strategies only see their sleeve's tickers
- [ ] Sum of all leaf targets == root investable budget (extension of the construct-to-budget invariant)
- [ ] Optimizer can tune sleeve-specific strategy weights independently
- [ ] Trade attribution carries sleeve path so reports can show "tech_concentration generated 47 trades" alongside "Momentum generated 47 trades"

## Out of scope

- Cross-sleeve correlation handling (a future enhancement; nice-to-have)
- Dynamic sleeve rebalancing (sleeve budgets that adapt to market regime — way out of scope)
- UI / visualization of the tree (separate concern)
- Per-sleeve cash reserves (only the root carries `min_cash_pct` — sleeves are fully invested within their budget)

## Related

- Depends on #32 (pluggable weighting policies) — every tree node uses a `WeightingPolicy`
- Depends on #29 (softmax allocator) — leaf nodes use the construct-to-budget primitive
- Related to #26 (strategy tier taxonomy) — the strategy tier redesign should land before sleeve-scoped strategies, otherwise we're wiring the new tree to the wrong strategy interface

**Comment by jugegp16** [comment]
Will add per ticker strategies later on, but this spec is not what we will do.
